### PR TITLE
support devfile components

### LINKED ISSUE #228: Verify odo 2.0 for s2i components
[kind/Enhancement]
## Which functionality do you think we should update/improve?

To make sure plugin is backward compatible the S2I components after switching to odo 2.0

### PR DESCRIPTION
**Pull request by sbouchet**
Signed-off-by: Stephane Bouchet <sbouchet@redhat.com>

## What is the purpose of this change? What does it change?
DevFile support with odo 2.0.0

## Was the change discussed in an issue?
fixes #228, #229

### CODE REVIEW

**Comment by sbouchet** [comment]
odo 2.0.1 introduce regression, so i still stick to 2.0.0 (https://github.com/openshift/odo/issues/4144)

**Comment by sbouchet** [comment]
back to draft as undeploy devfile component fails, similar to https://github.com/redhat-developer/vscode-openshift-tools/issues/1820

**Comment by jeffmaury** [comment]
I have the following setup:

- a component is deployed on my cluster
- I opened a project which as no .odo folder
- the component is displayed in the tree but with state **Pushed** but should be **No context**
- checked with master and it is correctly displayed as **No context**

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java [hunk 1]
```diff
@@ -104,641 +105,695 @@
...
+        execute(command, envVars, "logout");
+    }
+
+    private static List<Application> parseApplications(String json) throws IOException {
+        JSonParser parser = new JSonParser(JSON_MAPPER.readTree(json));
+        return parser.parseApplications();
+    }
+
+    @Override
+    public List<Application> getApplications(String project) throws IOException {
```
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java:460]: @jeffmaury @dgolovin : this temporary workaround is to bypass odo issues for application lists with devfiles

File: src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliApplicationTest.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.junit.Assert.*;
```
Comment (jeffmaury) [inline src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliApplicationTest.java:21]: Prefer named imports

File: src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliComponentTest.java [hunk 3]
```diff
@@ -65,7 +77,7 @@ public void checkCreateAndDiscoverComponent() throws IOException, InterruptedExc
         String application = APPLICATION_PREFIX + random.nextInt();
         String component = COMPONENT_PREFIX + random.nextInt();
         try {
-            createComponent(project, application, component, push);
+            createS2iComponent(project, application, component, push);
```
Comment (jeffmaury) [inline src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliComponentTest.java:80]: test method name should be renamed as well

File: src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliComponentTest.java [hunk 4]
```diff
@@ -86,8 +98,8 @@ public void checkCreateAndDeleteComponent() throws IOException, InterruptedExcep
         String application = APPLICATION_PREFIX + random.nextInt();
         String component = COMPONENT_PREFIX + random.nextInt();
         try {
-            createComponent(project, application, component, push);
-            odo.deleteComponent(project, application, COMPONENT_PATH, component, push);
+            createS2iComponent(project, application, component, push);
```
Comment (jeffmaury) [inline src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliComponentTest.java:101]: test method name should be renamed as well

File: src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliComponentTest.java [hunk 5]
```diff
@@ -151,16 +163,19 @@ public void checkCreateComponentAndCreateAndDeleteSecureURL() throws IOException
...
     public void checkCreateComponentAndLinkService() throws IOException, InterruptedException {
         String project = PROJECT_PREFIX + random.nextInt();
         String application = APPLICATION_PREFIX + random.nextInt();
         String component = COMPONENT_PREFIX + random.nextInt();
         String service = SERVICE_PREFIX + random.nextInt();
         try {
-            createComponent(project, application, component, push);
-            odo.createService(project, application, "postgresql-persistent", "default", service);
-            odo.link(project, application, component, COMPONENT_PATH, service, null);
+            createS2iComponent(project, application, component, push);
```
Comment (jeffmaury) [inline src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliComponentTest.java:172]: test method name should be renamed as well

File: src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliComponentTest.java [hunk 6]
```diff
@@ -205,7 +220,7 @@ public void checkCreateComponentAndListURLs() throws IOException, InterruptedExc
         String application = APPLICATION_PREFIX + random.nextInt();
         String component = COMPONENT_PREFIX + random.nextInt();
         try {
-            createComponent(project, application, component, push);
+            createS2iComponent(project, application, component, push);
```
Comment (jeffmaury) [inline src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliComponentTest.java:223]: test method name should be renamed as well

File: src/main/java/org/jboss/tools/intellij/openshift/actions/component/DebugComponentAction.java [hunk 7]
```diff
@@ -191,7 +186,7 @@ private boolean checkOdoDebugRunning(Odo odo, String projectName, String applica
             try {
                 // find an available port and use it
                 ServerSocket serverSocket = new ServerSocket(0);
-                port = serverSocket.getLocalPort();
+                port = Integer.valueOf(serverSocket.getLocalPort());
```
Comment (jeffmaury) [inline src/main/java/org/jboss/tools/intellij/openshift/actions/component/DebugComponentAction.java:189]: I don't see the rationale for this change ?

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/Component.java [hunk 8]
```diff
@@ -10,64 +10,85 @@
...
+    ComponentInfo getInfo();
+
+    ComponentKind getComponentKind();
+
+    class ComponentImpl implements Component {
+        private final String name;
+        private ComponentState state;
+        private String path;
+        private final ComponentInfo info;
+        private final ComponentKind kind;
```
Comment (jeffmaury) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/odo/Component.java:37]: Should be part of ComponentInfo

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/ComponentSourceType.java [hunk 9]
```diff
@@ -29,11 +27,14 @@ public String toString() {
     }
 
     public static ComponentSourceType fromAnnotation(String annotation) {
-        for(ComponentSourceType sourceType : values()) {
-            if (sourceType.annotation.equals(annotation)) {
-                return sourceType;
-            }
+        switch (annotation) {
```
Comment (jeffmaury) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/odo/ComponentSourceType.java:30]: I think the previous pattern was more extensible, what's the rationale for this change ?

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/Odo.java [hunk 10]
```diff
@@ -58,16 +58,16 @@
...
     void createURL(String project, String application, String context, String component, String name, Integer port,
                    boolean secure) throws IOException;
 
     void deleteURL(String project, String application, String context, String component, String name) throws IOException;
 
-    void undeployComponent(String project, String application, String context, String component) throws IOException;
+    void undeployComponent(String project, String application, String context, String component, ComponentKind kind) throws IOException;
 
-    void deleteComponent(String project, String application, String context, String component, boolean undeploy) throws IOException;
+    void deleteComponent(String project, String application, String context, String component, ComponentKind kind) throws IOException;
```
Comment (jeffmaury) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/odo/Odo.java:70]: the removal of the undeploy parameter is not clear to me, needs to discuss it

**Review by jeffmaury — CHANGES_REQUESTED** [review]

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java [hunk 11]
```diff
@@ -104,641 +105,695 @@
...
+        args.add(project);
+        args.add("--app");
+        args.add(application);
+        if (StringUtils.isNotBlank(componentVersion)) {
+            args.add("--s2i");
+        }
+        if (push) {
+            args.add("--now");
+        }
+        ExecHelper.executeWithTerminal(new File(source), args.toArray(new String[0]));
```
Comment (jeffmaury) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java:213]: this is wrong: you need tp pass envVars to support proxy

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java (hunk 1 above)
Comment (dgolovin) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java:460]: @sbouchet :+1:

File: src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliApplicationTest.java (hunk 2 above)
Comment (sbouchet) [reply src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliApplicationTest.java]: fixed.

File: src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliComponentTest.java (hunk 3 above)
Comment (sbouchet) [reply src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliComponentTest.java]: i created a new set of parameters to test both S2i and Devfile components.

File: src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliComponentTest.java (hunk 4 above)
Comment (sbouchet) [reply src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliComponentTest.java]: i created a new set of parameters to test both S2i and Devfile components.

File: src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliComponentTest.java (hunk 5 above)
Comment (sbouchet) [reply src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliComponentTest.java]: i created a new set of parameters to test both S2i and Devfile components.

File: src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliComponentTest.java (hunk 6 above)
Comment (sbouchet) [reply src/it/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCliComponentTest.java]: i created a new set of parameters to test both S2i and Devfile components.

File: src/main/java/org/jboss/tools/intellij/openshift/actions/component/DebugComponentAction.java (hunk 7 above)
Comment (sbouchet) [reply src/main/java/org/jboss/tools/intellij/openshift/actions/component/DebugComponentAction.java]: IntelliJ quick fix proposal. reverted.

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/Component.java (hunk 8 above)
Comment (sbouchet) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/odo/Component.java]: refactored to include kind in componentInfo.

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/Odo.java (hunk 10 above)
Comment (sbouchet) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/odo/Odo.java:70]: undeploy paramter is never used in the odoCli current implementation. 
tested with various states (pushed, no context, not pushed ) and both types (s2i/devfile )

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java (hunk 11 above)
Comment (sbouchet) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java]: fixed, and also removed method without envVars, not used anymore.

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/ComponentSourceType.java (hunk 9 above)
Comment (sbouchet) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/odo/ComponentSourceType.java]: I did some extra refactoring while creating ComponentKind enum.. reverted.

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java (hunk 1 above)
Comment (dgolovin) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java:460]: @sbouchet this approach can be used to build application view model when no cluster present. Currently view shows nothing when not connected and logged into the cluster.

Comment (sbouchet) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java:460]: yes, if odo can get the list of apps without being logged in, this can be a good enhancement.

Comment (sbouchet) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java:460]: @dgolovin : just discovered that workaround actually does not work as expected, because it only list apps for the current project set in the kubeconfig. 
Meaning that if there is multiple projects in the cluster, only the apps from the current one will be listed by this command... 
so we're still stuck and have to wait for odo to fix issues. 
`      --all-apps         List all components from all applications for the current set project`

**Comment by sbouchet** [comment]
@jeffmaury : can you review? unit test failure is tracked here: #244

**Comment by jeffmaury** [comment]
LGTM but faces the following issues:

- can't debug (but seems to be addressed in #231)
- can't create URL (but seems to be addressed in #123)
- debug menu is not available for Devfile Java components but it is for Devfile nodejs

**Comment by sbouchet** [comment]
Merging, the above comments will be addressed in separate issues.
